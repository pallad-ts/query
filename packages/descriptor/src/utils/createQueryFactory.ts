import { z } from "zod";
import { fromNullable } from "@sweet-monads/maybe";
import { QueryDescriptor } from "../QueryDescriptor";
import { fromZodValidation } from "@pallad/violations-zod";
import { right } from "@sweet-monads/either";

export function createQueryFactory(
	config: QueryDescriptor.Config,
	filtersValidator: QueryDescriptor.FiltersValidator<any> | undefined
) {
	const paginationSchema = getPaginationSchema(config.pagination);
	const sortingSchema = getSortingSchema(config.sorting);

	let schema = z.object({
		filters: z
			.object({})
			.passthrough()
			.default(() => ({})),
	});

	schema = paginationSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);
	schema = sortingSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);

	return (input: unknown) => {
		return fromZodValidation(schema, input).chain(({ filters, ...meta }) => {
			if (filtersValidator) {
				return filtersValidator(filters).map(filters => ({ ...meta, filters }));
			}
			return right({ ...meta, filters });
		});
	};
}

function createSortableFieldSchema(allowedFields: readonly [string, ...string[]]) {
	return z.object({
		field: z.enum(allowedFields),
		direction: z.enum(["DESC", "ASC"]),
	});
}

function getPaginationSchema(pagination: QueryDescriptor.Pagination | undefined) {
	return fromNullable(pagination).map(x => {
		if (x.type === "byCursor") {
			return z.object({
				limit: getLimitSchema(x),
				after: z.string().optional(),
				before: z.string().optional(),
			});
		}
		return z.object({
			limit: getLimitSchema(x),
			offset: z.number().int().min(0).default(0),
		});
	});
}

function getLimitSchema(pagination: QueryDescriptor.Pagination) {
	return z.number().int().min(1).max(pagination.maxLimit).default(pagination.defaultLimit);
}

function getSortingSchema(sorting: QueryDescriptor.Sorting<any> | undefined) {
	return fromNullable(sorting).map(x => {
		const sortByType = createSortableFieldSchema(x.sortableFields);
		if (x.type === "single") {
			return z.object({
				sortBy: sortByType.default(x.defaultSorting),
			});
		}

		return z.object({
			sortBy: z
				.array(sortByType)
				.nonempty()
				.default(x.defaultSorting as [any, ...any[]]),
		});
	});
}
