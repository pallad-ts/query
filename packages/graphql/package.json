{
  "name": "@pallad/query-graphql",
  "version": "2.0.1",
  "description": "Helper for graphql to create @pallad/query related types",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "scripts": {
    "test": "pallad-scripts test",
    "lint": "pallad-scripts lint",
    "compile": "pallad-scripts compile",
    "prepublishOnly": "npm run compile"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pallad-ts/query.git",
    "directory": "packages/main"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl/)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pallad-ts/query/issues"
  },
  "homepage": "https://github.com/pallad-ts/query#readme",
  "peerDependencies": {
    "graphql": "^16.3.0",
    "graphql-compose": "^9.0.8"
  },
  "dependencies": {
    "@pallad/builder": "^1.0.1",
    "@pallad/query": "^2.0.0",
    "predicates": "^2.0.3"
  }
}
