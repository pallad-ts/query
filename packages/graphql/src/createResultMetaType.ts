import {ObjectTypeComposerFieldConfigMapDefinition} from "graphql-compose/lib/ObjectTypeComposer";
import {PaginableByCursor, PaginableByOffset} from "@pallad/query";
import {ObjectTypeComposer} from "graphql-compose";
import {NonNullComposer} from "graphql-compose/lib/NonNullComposer";

export function createResultMetaType({baseName, paginationFields, sortType, extraMetaFields}: createResultMetaType.Options) {
	const metaFields: ObjectTypeComposerFieldConfigMapDefinition<any, any> = {};
	if (paginationFields) {
		Object.assign(metaFields, paginationFields);
	}

	if (sortType) {
		Object.assign(metaFields, {
			sortBy: {type: sortType}
		})
	}

	if (extraMetaFields) {
		Object.assign(metaFields, extraMetaFields);
	}

	if (!Object.keys(metaFields).length) {
		return;
	}

	return ObjectTypeComposer.createTemp({
		name: `${baseName}_Result_Meta`,
		fields: metaFields
	}).NonNull;
}

export namespace createResultMetaType {
	export interface Options {
		baseName: string;
		paginationFields?: ObjectTypeComposerFieldConfigMapDefinition<PaginableByCursor.ResultMeta | PaginableByOffset.ResultMeta, unknown>,
		sortType?: ObjectTypeComposer | NonNullComposer;
		extraMetaFields?: ObjectTypeComposerFieldConfigMapDefinition<unknown, unknown>
	}
}
