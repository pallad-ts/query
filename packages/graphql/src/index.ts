export * from './types';
export * from './createInputSortType';
export * from './createResultListType';
export * from './createSortFieldType';
export * from './createSortTypeBase';
export * from './GraphQLQueryBuilder';
