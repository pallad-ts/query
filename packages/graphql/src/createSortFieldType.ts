import {GraphQLEnumType} from "graphql";
import * as is from 'predicates'
import {camelCase} from "camel-case";

const assertSortableFields = is.assert(is.notEmptyArr);

/**
 * Creates enum for sortable fields
 */
export function createSortFieldType(baseName: string, sortableFields: readonly string[]) {
	assertSortableFields(sortableFields);
	return new GraphQLEnumType({
		name: `${baseName}_Sort_Field`,
		values: Object.fromEntries(
			sortableFields.map(x => {
				return [camelCase(x), {value: x}]
			})
		)
	})
}
