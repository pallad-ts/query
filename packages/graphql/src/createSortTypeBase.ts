import {ObjectTypeComposer} from "graphql-compose";
import {GraphQLNonNull, GraphQLEnumType} from "graphql";
import {GraphQLSortDirection} from "./types";

export function createSortTypeBase(name: string, sortFieldType: GraphQLEnumType) {
	return ObjectTypeComposer.createTemp({
		name: name,
		fields: {
			direction: {type: new GraphQLNonNull(GraphQLSortDirection)},
			field: {
				type: new GraphQLNonNull(
					sortFieldType
				)
			}
		}
	})
}
