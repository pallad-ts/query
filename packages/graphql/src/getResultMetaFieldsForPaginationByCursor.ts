import { GraphQLNonNull, GraphQLString } from "graphql";
import {GraphQLNonNegativeInt, GraphQLPositiveInt} from "graphql-scalars";
import { ObjectTypeComposerFieldConfigMapDefinition } from "graphql-compose/lib/ObjectTypeComposer";
import { PaginableByCursor } from "@pallad/query";

export function getResultMetaFieldsForPaginationByCursor(): ObjectTypeComposerFieldConfigMapDefinition<
	PaginableByCursor.ResultMeta,
	any
> {
	return {
		nextPage: { type: GraphQLString },
		previousPage: { type: GraphQLString },
		limit: { type: new GraphQLNonNull(GraphQLNonNegativeInt) },
	};
}
