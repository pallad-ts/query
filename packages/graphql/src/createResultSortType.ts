import {GraphQLEnumType} from "graphql";
import {createSortTypeBase} from "./createSortTypeBase";

export function createResultSortType({baseName, sortFieldType, isMulti}: createResultSortType.Options) {
	const type = createSortTypeBase(`${baseName}_Result_Sort`, sortFieldType);
	if (isMulti) {
		return type.NonNull.List.NonNull;
	}
	return type.NonNull;
}

export namespace createResultSortType {
	export interface Options {
		baseName: string,
		sortFieldType: GraphQLEnumType,
		/**
		 * Whether multiple sorting fields are allowed
		 */
		isMulti: boolean;
	}
}
