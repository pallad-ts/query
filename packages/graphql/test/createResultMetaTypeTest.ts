import {createResultMetaType} from "@src/createResultMetaType";
import {NonNullComposer} from "graphql-compose";
import {printObject} from "graphql-compose/lib/utils/schemaPrinter";
import {GraphQLObjectType} from "graphql/type";
import {createSortTypeBase} from "@src/createSortTypeBase";
import {createResultSortType} from "@src/createResultSortType";
import {createSortFieldType} from "@src/createSortFieldType";
import {SORTABLE_FIELDS} from "./fixtures";
import {getResultMetaFieldsForPaginationByCursor} from "@src/getResultMetaFieldsForPaginationByCursor";
import {GraphQLString} from "graphql";


describe('createResultMetaType', () => {
	const sortType = createResultSortType({
		baseName: 'Base',
		isMulti: true,
		sortFieldType: createSortFieldType('Base', SORTABLE_FIELDS)
	});

	it('returns undefined if no meta fields computed', () => {
		const type = createResultMetaType({
			baseName: 'Base'
		});

		expect(type).toBeUndefined();
	});

	it.each<[string, Omit<createResultMetaType.Options, 'baseName'>]>([
		[
			'with sort type',
			{
				sortType: sortType!
			}
		],
		[
			'with pagination fields',
			{
				paginationFields: getResultMetaFieldsForPaginationByCursor()
			}
		],
		[
			'with extra meta fields',
			{
				extraMetaFields: {
					name: {type: GraphQLString}
				}
			}
		],
		[
			'with all options',
			{
				sortType: sortType!,
				paginationFields: getResultMetaFieldsForPaginationByCursor(),
				extraMetaFields: {
					name: {type: GraphQLString}
				}
			}
		]
	])('creates result meta type: %s', (_, opts) => {
		const type = createResultMetaType({
			baseName: 'Base',
			...opts
		});

		expect(type)
			.toBeInstanceOf(NonNullComposer);

		expect(
			printObject(type!.getUnwrappedTC().getType() as GraphQLObjectType)
		)
			.toMatchSnapshot();
	});
});
