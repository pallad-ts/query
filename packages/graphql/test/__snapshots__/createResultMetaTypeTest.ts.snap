// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResultMetaType creates result meta type: with all options 1`] = `
"type Base_Result_Meta {
  nextPage: String
  previousPage: String
  limit: PositiveInt!
  sortBy: [Base_Result_Sort!]!
  name: String
}"
`;

exports[`createResultMetaType creates result meta type: with extra meta fields 1`] = `
"type Base_Result_Meta {
  name: String
}"
`;

exports[`createResultMetaType creates result meta type: with pagination fields 1`] = `
"type Base_Result_Meta {
  nextPage: String
  previousPage: String
  limit: PositiveInt!
}"
`;

exports[`createResultMetaType creates result meta type: with sort type 1`] = `
"type Base_Result_Meta {
  sortBy: [Base_Result_Sort!]!
}"
`;
